rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+    [strongly typed](https://github.com/google/ml-metadata/blob/ba69ae039bd2205ec2d7b982b3bfdda4718bf8df/ml_metadata/proto/metadata_store.proto#L55)
+    [artifact metadata](https://github.com/google/ml-metadata/blob/ba69ae039bd2205ec2d7b982b3bfdda4718bf8df/ml_metadata/proto/metadata_store.proto#L29).
+    This is a key value proposition of TFX DSL.
+
+    *   As of [kfp 0.1.26](https://github.com/kubeflow/pipelines/tree/0.1.26),
+        the
+        [KFP DSL](https://github.com/kubeflow/pipelines/tree/0.1.26/sdk/python/kfp/dsl)
+        has type system to input- and output- of a component. However, it is in
+        a way that it doesn’t attach semantic meaning to the input- and output-
+        variables in a way compatible to ML Metadata, and the way how TFX
+        uitilizes it to realize its functionalies.
+
+*   **Metadata-driven component execution**: Metadata-centric interface of
+    components enables TFX Driver and Publisher logic to be applied
+    consistently, thus enabling caching of component execution as well
+    component-specific (meta)data-driven execution decisions
+
+    *   *Example*: ModelValidator can validate models trained with *any*
+        component, so long as the produced artifact is of the *type* Model.
+
+*   **Inter-Component communication via ML Metadata store**: This enables higher
+    order component-specific driver logic that depends on another component’s
+    behavior.
+
+    *   *Example*: Pusher can choose to proceed or halt pushing depending on
+        output artifact from ModelValidator.
+
+*   **Ability to share and reuse a wrapped component as drop-in replacement with
+    another**: As a result of the artifact centric, strongly typed input- and
+    output- signatures, it enables robust sharing and drop-in replacement of
+    components, so long as signatures (the list of input- and output- artifact
+    *types*) are the same.
+
+Additionally, the proposed container-based component will enable the following
+new features in TFX DSL, which already exist In Kubeflow Pipelines:
+
+*   **Ability to use any containerized application program in a pipeline**: The
+    proposed container-based component does not preclude any containerized
+    application programs from being used as a pipeline step.
+
+*   **Ability to have fine-grained control on the underlying k8s execution
+    environment**: The proposed container-based component preserves the user’s
+    ability to control underlying Kubernetes runtime configuration for
+    containerized program execution.
+
+## Design Proposal
+
+As stated previously, custom containers may be written in arbitrary languages.
+The input interface to containers is restricted to the command-line used to
+invoke the application, while the output interface is through file I/O (either
+STDOUT, STDOUT, or container-local files).
+
+Since container applications may not have access to TFX libraries, they are not
+able to (or don’t even wish to) serialize/de-serialize
```

Comment:
Do you mean a case where some upstream component would select a particular instance of `HParams`, and the downstream component to consume it as execution property?

If so, because we model any output from component as an artifact, the current way of thinking is to define an `HParams` artifact as the output from the upstream component, and materialize it as an artifact, and have the downstream component consume it as one of the input artifacts. Doing so has side benefit to materialize the selected `HParams` in Metadata store.

Having said that, we are open to suggestion if there would be an alternative.

\cc @neuromage

Reply:
For hparams, that makes sense. An alternative here could be add mlmd::Execution as an output from component, which includes the status (e.g., success/fail/aborted/cancel ..) of the execution and extra/arbitrary run time properties of the execution, the orchestrator runtime reads this Execution, and update its record in mlmd too.

Reply:
@hughmiao this is a nice idea. Using `mlmd::Execution` seems like a nice structured approach to parsing returned data. As for the mechanism, I'm still leaning towards file-based approach, which scales to different executors.